modules/table/src/main/java/org/apache/ignite/internal/table/distributed/raft/snapshot/FullStateTransferIndexChooser.java

```diff
@@ -274,30 +267,58 @@ private long catalogActivationTimestampBusy(int catalogVersion) {
         return catalog.time();
     }
 
-    private void recoveryReadOnlyIndexesBusy() {
+    private void recoverReadOnlyIndexesBusy() {
         int earliestCatalogVersion = catalogService.earliestCatalogVersion();
         int latestCatalogVersion = catalogService.latestCatalogVersion();
 
-        // At the moment, we will only use tables from the latest version (not dropped), since so far only replicas for them are started
-        // on the node.
-        int[] tableIds = catalogService.tables(latestCatalogVersion).stream().mapToInt(CatalogObjectDescriptor::id).toArray();
-
-        Map<Integer, ReadOnlyIndexInfo> readOnlyIndexes = new HashMap<>();
+        var readOnlyIndexById = new HashMap<Integer, ReadOnlyIndexInfo>();
+        var previousCatalogVersionTableIds = Set.<Integer>of();
 
         for (int catalogVersion = earliestCatalogVersion; catalogVersion <= latestCatalogVersion; catalogVersion++) {
-            Catalog catalog = catalogService.catalog(catalogVersion);
+            long activationTs = catalogActivationTimestampBusy(catalogVersion);
 
-            assert catalog != null : catalogVersion;
+            catalogService.indexes(catalogVersion).stream()
+                    .filter(index -> index.status() == STOPPING)
+                    .forEach(index -> readOnlyIndexById.computeIfAbsent(index.id(), i -> new ReadOnlyIndexInfo(index, activationTs)));
```

Comment:
This might become incorrect if, in future, Catalog compaction will be enabled. We might find non-first version in which the index is STOPPING, because first version where it became STOPPING is already removed. Do we need a TODO about this?

Reply:
Add todos